data/scripts.js

```diff
@@ -292,11 +255,291 @@ let BattleScripts = {
+			// 4. check for prankster immunity
+			this.pranksterImmunity,
+
+			// 5. check accuracy
+			this.accuracy,
+
+			// 6. break protection effects
+			this.breakProtect,
+
+			// 7. steal positive boosts (Spectral Thief)
+			this.stealBoosts,
+
+			// 8. loop that processes each hit of the move (has its own steps per iteration)
+			this.moveHitLoop,
+
+			// 9. effects that run after secondary effects, such as Color Change
+			this.afterMoveSecondaryEvent,
+		];
+		if (this.gen >= 7) {
+			// Swap step 1 with step 2
+			moveSteps.splice(2, 0, moveSteps.splice(1, 1)[0]);
+		}
+
+		this.setActiveMove(move, pokemon, targets[0]);
+		move.zBrokeProtect = false;
+
+		let hitResult = this.singleEvent('PrepareHit', move, {}, targets[0], pokemon, move);
+		if (!hitResult) {
+			if (hitResult === false) {
+				this.add('-fail', pokemon);
+				this.attrLastMove('[still]');
+			}
+			return false;
+		}
+		this.runEvent('PrepareHit', pokemon, targets[0], move);
+
+		if (!this.singleEvent('Try', move, null, pokemon, targets[0], move)) {
+			return false;
+		}
+
+		let finalResult;
+		let hitResults;
+		for (const step of moveSteps) {
+			hitResults = step.call(this, targets, pokemon, move);
+			if (!hitResults) continue;
+			for (let i = 0; i < targets.length; i++) {
+				if (!hitResults[i]) {
+					// hit resolved for this target.
+					// remove target and corresponding hitResults entry from their respective arrays
+					targets.splice(i, 1);
+					let targetResult = hitResults.splice(i, 1)[0];
+					i--;
+					// store failure/NOT_FAILURE in finalResult, with the following priority:
+					// truthy/number > false > '' > undefined
+					if (targetResult || targetResult === 0) {
+						finalResult = targetResult;
+					} else if (finalResult !== false || targetResult !== this.NOT_FAILURE) {
+						finalResult = targetResult;
+					}
+				}
+			}
+		}
+
+		let moveResult = !!targets.length;
+		if (!moveResult && finalResult === this.NOT_FAILURE) pokemon.moveThisTurnResult = null;
+		if (move.spreadHit) this.attrLastMove('[spread] ' + targets.join(','));
+		return moveResult;
+	},
+	tryImmunityEvent(targets, pokemon, move) {
+		let hitResults = this.runEvent('TryImmunity', targets, pokemon, move);
+		targets.forEach((target, index) => {
+			if (hitResults[index] === false) {
+				if (!move.spreadHit) this.attrLastMove('[miss]');
+				this.add('-miss', pokemon, target);
+			} else {
+				hitResults[index] = false;
+			}
+		});
+		return hitResults;
+	},
+	typeImmunity(targets, pokemon, move) {
+		if (move.ignoreImmunity === undefined) {
+			move.ignoreImmunity = (move.category === 'Status');
+		}
+
+		let hitResults = [];
+		for (let i = 0; i < targets.length; i++) {
+			hitResults[i] = (move.ignoreImmunity && (move.ignoreImmunity === true || move.ignoreImmunity[move.type])) || targets[i].runImmunity(move.type, true);
+		}
+
+		return hitResults;
+	},
+	tryHitEvent(targets, pokemon, move) {
+		let hitResults = this.runEvent('TryHit', targets, pokemon, move);
```

Comment:
Ok, new problem: this is somehow not passing an array.

Reply:
"not passing an array"?

Reply:
To test this, I put a `console.log('' + target);` inside `runEvent` and it was almost always a single Pokemon despite TypeScript being satisfied that this is always an array.

Reply:
Oh, um, I don't think an array is a valid target for `runEvent`. It only takes single pokemon, sides, or battles, right?

Reply:
Yeah, `target` must be `Pokemon | Side | Battle | null`. `targets` is an array, right? `Pokemon[]`? That's not a valid event target.

Reply:
In theory TypeScript should be yelling at you for that.

Reply:
I modified `runEvent` to accept and handle arrays, but it appears to not even be receiving them.

Reply:
Ok, I figured it out. The problem is that `Pokemon.getMoveTargets`, which is supposed to only return a `Pokemon` array, isn't returning an array.
`
let targets = pokemon.getMoveTargets(move, target);
		console.log('1. ' + targets instanceof Array);
`
^This prints `false`.

Reply:
...because order of operations was wrong. Adding some parens makes it print `1. true`.

Reply:
Why did you modify `runEvent` to handle arrays? Is there a reason why you can't run each target individually?

Reply:
Because the `TryHit` event needs to proc its effects in an order that respects `onTryHitPriority` even with multiple targets. That is, Protect always triggers before Lightning Rod, instead of just going left-to-right like how we'd have it, so we need to gather _all_ event handlers from all the targets and sort them together. Additionally, the `AfterMoveSecondary` event even respects speed order instead of the left-to-right order that all the other events use.

Reply:
Oh. :(